.github/workflows/linux-general.yml

```diff
@@ -0,0 +1,359 @@
+        id: cache
+        continue-on-error: true
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.CACHE_DIR }}
+          key: ${{ runner.os }}-nih-${{ steps.cachekeys.outputs.cmakeconfig }}-${{ matrix.image }}-${{ matrix.ninja_ccs }}-${{ matrix.make_ccs }}-${{ steps.cachekeys.outputs.instructions }}
+
+      - name: Initialize CMake cache
+        if: ${{ ! steps.cache.outputs.cache-hit }}
+        shell: bash
+        run: |
+            if [[ -e "${NIH_CACHE_ROOT}" ]]
+            then
+                echo Cache is already set up.
+                exit
+            fi
+
+            cmake --version
+
+            if [ "${CMAKE_ADD}" != "" ]; then
+                export CMAKE_EXTRA_ARGS+="${CMAKE_ADD}"
+            fi
+
+            # Builds that use unix-makefiles
+            for CC in ${{ matrix.make_ccs }}
+            do
+                for BUILD_TYPE in ${{ matrix.make_types }}
+                do
+                    export CC BUILD_TYPE
+                    # Set CXX to the C++ complier corresponding to CC
+                    # It's a little hacky, but keeps this job's matrix
+                    #  a little simpler
+                    export CXX=${CC/gcc/g++}
+                    export CXX=${CXX/clang/clang++}
+
+                    ${CC} --version
+                    ${CXX} --version
+                    dir="build/$( basename ${CXX} ).${BUILD_TYPE}.makefile"
+                    mkdir "-pv" "${dir}"
+                    pushd "${dir}"
+                    set "-x"
+                    # Verbose or not, it's good to see the right params
+                    # are being used
+                    cmake -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+                        ${CMAKE_EXTRA_ARGS} ../..
+                    set +x
+                    popd
+                done
+            done
+
+            # Builds that use Ninja
+            export CMAKE_EXTRA_ARGS+=" -G Ninja"
+            for CC in ${{ matrix.ninja_ccs }}
+            do
+                for BUILD_TYPE in ${{ matrix.ninja_types }}
+                do
+                    export CC BUILD_TYPE
+                    # Set CXX to the C++ complier corresponding to CC
+                    # It's a little hacky, but keeps this job's matrix
+                    #  a little simpler
+                    export CXX=${CC/gcc/g++}
+                    export CXX=${CXX/clang/clang++}
+
+                    ${CC} --version
+                    ${CXX} --version
+                    dir="build/$( basename ${CXX} ).${BUILD_TYPE}"
+                    mkdir "-pv" "${dir}"
+                    pushd "${dir}"
+                    set "-x"
+                    # Verbose or not, it's good to see the right params
+                    # are being used
+                    cmake -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+                        ${CMAKE_EXTRA_ARGS} ../..
+                    set +x
+                    popd
+                done
+            done
```

Comment:
This bundles the generated build files for a bunch of different builds into a single cache. Could it be better to separate them, so that the cache is smaller, they can be built in parallel, etc.?

Reply:
If I separate them, I may as well get rid of this step altogether (I'm not against that) because the base job for each config will download the dependencies itself.

The reason for putting everything into a single cache is that, with the change to store all the source code in a top-level folder, the size of that source code dwarfs the size of the individual config folders.

The job outputs some of the cache folder sizes, for example at https://github.com/ripple/rippled/runs/3186329095?check_suite_focus=true#step:7:1054
```
200M	/__w/rippled/rippled/_cache/nih_c
4.6M	/__w/rippled/rippled/_cache/nih_c/ninja
1.2M	/__w/rippled/rippled/_cache/nih_c/ninja/GNU_9.2.1
1.2M	/__w/rippled/rippled/_cache/nih_c/ninja/Clang_9.0.1
1.2M	/__w/rippled/rippled/_cache/nih_c/ninja/GNU_8.3.0
1.2M	/__w/rippled/rippled/_cache/nih_c/ninja/Clang_8.0.1
676K	/__w/rippled/rippled/_cache/nih_c/unix_makefiles
672K	/__w/rippled/rippled/_cache/nih_c/unix_makefiles/GNU_8.3.0
```
It skips over the `src` folder (derp), but you can see the total size is 200M, while the total of the job-specific folders are about 5M. The rest is `src`.

Now, all that said, I realize that this `src` folder ends up getting duplicated across *all* of the jobs anyway, because they all use this cache as a "seed" for their own cache. Thinking about it now, I think I could make it significantly better:
1. Download `src` into a separate cache, and have the other jobs use that as an immutable dependency. This will be similar to the Windows jobs, which build vcpkg and boost dependencies once, and share them among all the other jobs. Assuming it works, this will take less space, and a little less time, because I'd only need to call `cmake` once.
2. Get rid of this cache step entirely and let the jobs download their own dependencies.

I'll play around with this a little bit and see what I can figure out.

Reply:
*one month later.gif*

It took a while due to other priorities, and the fact that test iteration can take a while on CI, but I've reorganized the caches sort of as described in option 1 for the Linux and Windows jobs. With a little experimentation, I modified the "seed cache" jobs to build the dependencies **once per docker image**, which seems to be the only way to force them to download themselves and their dependencies. It also preserves the "stamp" folders, which prevents the jobs from re-downloading. The stamp folders are compatible (enough) that they can be used by the different compilers. The build output is not, so I just throw it away. Not perfect, but a significant improvement. I haven't updated the MacOS jobs yet, because I wanted to get this pushed now that it's stablized.

One issue causing delays is that I was having consistent failures in the Windows jobs that all seemed to come back to the `Check` tests, but I can't find anything wrong with those tests. I'm pretty sure it's resource related, but rather than spend another month tracking it down remotely, I decided that this problem is out of scope for this PR, and punted, adding an option to make those tests (or any other tests that have similar problems down the road) manual at build time, then have CI run them manually. That's the `[FOLD] Make the check tests manual on Github CI Windows` commit.